### ISSUE TITLE
Memory leak on Linux

### ISSUE THREAD

**Issue by deadman2000** [SciSharp/TensorFlow.NET#418]
Session `close` calling not released resources and tf threads are not stopped.

Example project demonstrated threads count increase: https://github.com/deadman2000/TFNetMemoryLeak

It's not TF bug, i tested on similar project on C: https://github.com/deadman2000/TFCThreadTest

It's Linux only problem. On Windows resources released success

**Comment by Oceania2018** [comment]
@deadman2000 You have to dispose Graph explictly.

**Comment by deadman2000** [comment]
Manually disposing Graph not helps
```cs
            session.graph.Dispose();
            session.close();
```

Test log:
```
2019-10-10 12:42:43.235019: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: model/
2019-10-10 12:42:43.238466: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { serve }
2019-10-10 12:42:43.249303: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.
2019-10-10 12:42:43.276624: I tensorflow/cc/saved_model/loader.cc:151] Running initialization op on SavedModel bundle at path: model/
2019-10-10 12:42:43.284454: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { serve }; Status: success. Took 49441 microseconds.
Dispose tf.Tensor '<unnamed Operation>' shape=(1,2) dtype=TF_FLOAT
Dispose Tensor disposing:True _disposed:False
  TF_DeleteTensor
Dispose grap-key-45/, (39260016)
Dispose Graph disposing:True _disposed:False
  TF_DeleteGraph
Dispose Tensorflow.Session
Dispose Session disposing:True _disposed:False
DisposeUnmanagedResources
  TF_DeleteSession
Dispose Tensorflow.Status
Dispose Status disposing:True _disposed:False
  TF_DeleteStatus
End
Threads: 71
Press Q to break or any another to repeat
2019-10-10 12:42:43.928763: I tensorflow/cc/saved_model/reader.cc:31] Reading SavedModel from: model/
2019-10-10 12:42:43.932001: I tensorflow/cc/saved_model/reader.cc:54] Reading meta graph with tags { serve }
2019-10-10 12:42:43.941843: I tensorflow/cc/saved_model/loader.cc:202] Restoring SavedModel bundle.
2019-10-10 12:42:43.969191: I tensorflow/cc/saved_model/loader.cc:151] Running initialization op on SavedModel bundle at path: model/
2019-10-10 12:42:43.973776: I tensorflow/cc/saved_model/loader.cc:311] SavedModel load for tags { serve }; Status: success. Took 45020 microseconds.
Dispose tf.Tensor '<unnamed Operation>' shape=(1,2) dtype=TF_FLOAT
Dispose Tensor disposing:True _disposed:False
  TF_DeleteTensor
Dispose grap-key-49/, (38424432)
Dispose Graph disposing:True _disposed:False
  TF_DeleteGraph
Dispose Tensorflow.Session
Dispose Session disposing:True _disposed:False
DisposeUnmanagedResources
  TF_DeleteSession
Dispose Tensorflow.Status
Dispose Status disposing:True _disposed:False
  TF_DeleteStatus
End
Threads: 74
Press Q to break or any another to repeat
```

**Comment by Nucs** [comment]
This might also leak in Linux for a different reason but for me on Windows.
Every time you run program (press any key in the console program in the repository) there is memory build up.

![image](https://user-images.githubusercontent.com/649919/66935683-18bd9c80-f045-11e9-94b7-28162325a8ca.png)

Only the following line causes leaking (with or without `.as_default()`.
```C#
var session = Session.LoadFromSavedModel(modelLocation);`
```

@Oceania2018 To your attention.

**Comment by gosha20777** [comment]
I have the same problem with Linux

**Comment by LukeBolly** [comment]
I'm having a similar problem, I've got about 20 unit tests and memory isn't being completely returned after each one. I'm calling:
_session.graph.Dispose();
_session.Dispose();

I'm seeing slight growth after each one that loads a saved model
![image](https://user-images.githubusercontent.com/24449147/121869243-6a478e80-cd34-11eb-98bd-beb14fa87c89.png)

I am also using
Session.LoadFromSavedModel(modelLocation);

**Comment by Oceania2018** [comment]
@gosha20777 @LukeBolly Could you PR an minimul runnable code into https://github.com/SciSharp/TensorFlow.NET/tree/master/src/TensorFlowNet.Benchmarks/Leak?

**Comment by LukeBolly** [comment]
@Oceania2018 I haven't had time to create a repro for you, but I've been debugging another issue and have come across some stuff leftover from TensorFlow.NET in a memory dump after my TF process has finished and calling dispose on the graph and session

![image](https://user-images.githubusercontent.com/24449147/130741929-cd5ebb57-16e9-47ea-b636-bbfa9b8815fa.png)

**Comment by LukeBolly** [comment]
Hi @Oceania2018, I've updated to the latest version. After running all of my unit tests and checking the managed memory at the end, there are still a large number of Tensorflow objects left in memory.

I've gone through my code and added Dispose() to all graphs, sessions, NDArrays, but I still end up with this locked up in memory:
![image](https://user-images.githubusercontent.com/24449147/131442807-c1832294-def5-47e8-8bb8-fa29a308917b.png)

**Comment by LukeBolly** [comment]
While I can't share the model I'm using, loading and disposing it in a loop confirms there is an issue cleaning up resources when working with a SavedModel.

```
public class TestModel
{
        private readonly Session _classifierSession;
        
        public TestModel(string classifierModelPath )
        {
	        for (var i = 0; i < 1000; i++)
	        {
		        _classifierSession = Session.LoadFromSavedModel(classifierModelPath );
        
		        _classifierSession.graph.Exit();
		        _classifierSession.graph.Dispose();
		        _classifierSession.Dispose();
	        }
        }
}
```

![image](https://user-images.githubusercontent.com/24449147/131465336-f3666a8e-d2a3-4e7e-a75d-8e918174d474.png)

**Comment by Oceania2018** [comment]
@LukeBolly Does it apply to any model? whether the other models cause the same issue?
Another way might be able to release the resource is `tf.Context.reset_context()`.

**Comment by LukeBolly** [comment]
Yep, I've run it again with a very simple model which is just a bunch of Conv layers and a call signature, and I'm seeing the same behavior:

![image](https://user-images.githubusercontent.com/24449147/131523491-194dffb9-870b-4052-8d53-46864e7a0c61.png)

This model is small and initialises much faster so the chart is smoother, but its the same behavior:
![image](https://user-images.githubusercontent.com/24449147/131523631-7246ac98-5deb-42e7-a8a1-db3a3e443c0b.png)

**Comment by LukeBolly** [comment]
I'll try to get a repro up for you tomorrow if I get time.

**Comment by LukeBolly** [comment]
@Oceania2018 
Unfortunately the fix added appears to bypass loading the graph entirely, see here:

![image](https://user-images.githubusercontent.com/24449147/132445678-6ed6becb-460d-4e25-9a39-437312e20154.png)

The graph is disposed as soon as it loads, as a result it isn't usable in any way.

**Comment by LukeBolly** [comment]
@Oceania2018 I've put up a PR that fixes the issue with the graph and extended the test to ensure it runs, it seems like there is still a small leak somewhere though: #858
